src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts

```diff
@@ -0,0 +1,685 @@
+        );
+        let refreshRequired = true;
+
+        switch (payload.command) {
+            case DataWranglerCommands.ExportToCsv:
+                await notebook?.execute(`${currentVariableName}.to_csv("./cleaned.csv", index=False)`, '', 0, uuid());
+                throw new Error('Not Implemented');
+                break;
+
+            case DataWranglerCommands.ExportToPythonScript:
+                await this.generatePythonCode();
+                break;
+
+            case DataWranglerCommands.ExportToNotebook:
+                await this.generateNotebook();
+                break;
+
+            case DataWranglerCommands.RenameColumn:
+                historyItem = this.renameColumn(currentVariableName, payload.args as IRenameColumnsRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.Drop:
+                historyItem = this.drop(currentVariableName, payload.args as IDropRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.DropDuplicates:
+                historyItem = this.dropDuplicates(currentVariableName, payload.args as IDropDuplicatesRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.DropNa:
+                historyItem = this.dropNa(currentVariableName, payload.args as IDropNaRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.PyplotHistogram:
+                refreshRequired = false;
+                code = `import matplotlib.pyplot as plt\nplt.hist(${currentVariableName}["${payload.args.target}"])\n`;
+                break;
+
+            case DataWranglerCommands.NormalizeColumn:
+                historyItem = this.normalizeColumn(currentVariableName, payload.args as INormalizeColumnRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.FillNa:
+                historyItem = this.fillNa(currentVariableName, payload.args as IFillNaRequest);
+                code = historyItem.code;
+                newVariableName = historyItem.variableName;
+                break;
+
+            case DataWranglerCommands.GetHistoryItem:
+                this.getHistoryItem(payload.args.index).ignoreErrors();
+                break;
+
+            case DataWranglerCommands.Describe:
+                void this.getColumnStats(payload.args.columnName);
+                break;
+        }
+
+        const dataCleaningMode = this.configService.getSettings().dataCleaningMode;
```

Comment:
You might abstract out this two pronged approach into separate classes that implement an interface? DataWranglerCommandHandler or something like that? One that executes in a notebook. One that is standalone?

Reply:
Or just two functions at least with the same signature.

Reply:
I think the commands could also be a map from command to function (instead of the big switch statement), each function returning a historyItem or undefined if no further history update required.

Reply:
I see what you mean with the separate classes implementing an interface. It would especially be smart once other forms are added like WITH_INTERACTIVE_WINDOW, etc. It started getting tricky due to the async/awaits and disposables. For the two functions, I believe that doesn't work in typescript because of it getting compiled to javascript (or I may be misinterpreting what you mean). I do like the command to function idea, but there are some async/await functions which makes it a little awkward. I can make them all async. Do you think that would be worthwhile exploring?

Reply:
Making them all async sounds fine to me. Non async functions that are marked that way don't infer a penalty AFAIK. They just run synchronously.